### PR TITLE
ADR for blob storage

### PR DESCRIPTION
**Pull request by arildbjerkenes**

### CODE REVIEW

File: doc/adr/0004-new-blob-storage.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# 4. New blob storage
+Date 2024-01-24
+
+## Status
+
+Accepted
+
+## Context
+
+In PCS 4 we store attachments as blobs in an Azure storage container with the following file structure: \<container\>/\<plant\>/\<Trunc(\<file_id\>/1000)*1000/\<file_id\>.jpg. The structure is based on how things where stored earlier, when the files where stored in folders, with 1000 files in each folder (based on a limit for folders).  To add a new blob, we use Oracle SEQUENCE to find the next id. This id is used to calculate what folder to use and the name on the file. There have been incidents in the system causing two attachments pointing to the same blob.
```
Comment (janid1967) [inline doc/adr/0004-new-blob-storage.md:10]: based on a limit in Operation System for folders

File: doc/adr/0004-new-blob-storage.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+For IPO and Preservation we have used a simpler structure, where we use Guid instead of id, and without needing to calculate the folder based on an id.  We have the following structure:  \<container\>/\<plant\>/\<objekttype\>/\<guid\>/bilde.jpg.
```
Comment (janid1967) [inline doc/adr/0004-new-blob-storage.md:12]: objecttype

File: doc/adr/0004-new-blob-storage.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+## Decision
+
+For Completion we will have a new storage container for attachment blobs, with the same file-structure as for IPO and Preservation. All blobs (for all entitites) in the old storage container will be moved to the new.
```
Comment (janid1967) [inline doc/adr/0004-new-blob-storage.md:16]: For Completion we will use the same storage container for attachment blobs, with the same ..

File: doc/adr/0004-new-blob-storage.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+For IPO and Preservation we have used a simpler structure, where we use Guid instead of id, and without needing to calculate the folder based on an id.  We have the following structure:  \<container\>/\<plant\>/\<objekttype\>/\<guid\>/bilde.jpg.
+
+## Decision
+
+For Completion we will have a new storage container for attachment blobs, with the same file-structure as for IPO and Preservation. All blobs (for all entitites) in the old storage container will be moved to the new. 
+
+## Consequences
+
+We need to copy all blobs from PCS 4 storage container to the new storage container for Completion, prior to going live with the first version of Completion. 
+To do this task we can use Azure Storage Data Movement Library for .Net".
```
Comment (janid1967) [inline doc/adr/0004-new-blob-storage.md:21]: Ahhh .. sure on that?
We need to perform custom code to calculate new folder/name of blob in new container.

**Comment by janid1967** [comment]
Open solution in VS and add the adr into solution items

**Review by janid1967 — CHANGES_REQUESTED** [review]
See comments

**Review by hnumequinor — APPROVED** [review]

File: doc/adr/0004-new-blob-storage.md (hunk 4 above)
Comment (arildbjerkenes) [reply doc/adr/0004-new-blob-storage.md]: Azure Storage Data Movement Library for .Net is a framework we you can copy a file from one storage container to another. I assume that we can set what folder to store the file in, and also change the file name. We will need to create our own 'tool', probably based on the one we used when we did M2C. By using this framework we might not need to download the file, and then upload it again. This ADR is not a decision on how to copy the files.

File: doc/adr/0004-new-blob-storage.md (hunk 3 above)
Comment (arildbjerkenes) [reply doc/adr/0004-new-blob-storage.md]: What do you mean by 'the same storage container'? The same as what? The decision we make is to have a new storage container for Completion. Isn't it?

**Review by janid1967 — APPROVED** [review]

**Pull request merged**